catalog/lease: use txn time when detecting session based leasing table

fqazi opened the pull request:
Previously, the logic to detect if the session based leasing table
should be injected synthetically queries the system database version,
but did not correctly use the transaction timestamp, which could lead to
us incorrectly assuming the descriptor is upgraded at the timestamp that
the leasing count query executed at. To address this, this patch updates
the CachedDatabaseRegions to support using a fixed timestamp from the
transaction.

Fixes: #120675, #121036

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/121070)
<!-- Reviewable:end -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121070)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/catalog/lease/count.go` line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121070#-NttAoab4KAHXke34zYH:-NttAoab4KAHXke34zYI:b-r0kx48) ([raw file](https://github.com/cockroachdb/cockroach/blob/4aad97ec0844e431c638b681ef50b594879e1c4e/pkg/sql/catalog/lease/count.go#L54)):*
> ```Go
> 	var systemDBVersion *roachpb.Version
> 	if cachedDatabaseRegions != nil {
> 		systemDBVersion = cachedDatabaseRegions.GetSystemDatabaseVersion()
> ```

could we avoid the retry if we made sure that the systemDBVersion here is read at an appropriate timestamp. i'm thinking of something like
```
diff --git a/pkg/sql/conn_executor_exec.go b/pkg/sql/conn_executor_exec.go
index 6b4352c25d1..3aa7f8cfc55 100644
--- a/pkg/sql/conn_executor_exec.go
+++ b/pkg/sql/conn_executor_exec.go
@@ -1417,7 +1417,7 @@ func (ex *connExecutor) checkDescriptorTwoVersionInvariant(ctx context.Context)
     if knobs := ex.server.cfg.SchemaChangerTestingKnobs; knobs != nil {
         inRetryBackoff = knobs.TwoVersionLeaseViolation
     }
-    regionCache, err := regions.NewCachedDatabaseRegions(ctx, ex.server.cfg.DB, ex.server.cfg.LeaseManager)
+    regionCache, err := regions.NewCachedDatabaseRegions(ctx, ex.server.cfg.DB, ex.server.cfg.LeaseManager, ex.state.mu.txn.ProvisionalCommitTimestamp())
     if err != nil {
         return err
     }
diff --git a/pkg/sql/regions/cached_db_regions.go b/pkg/sql/regions/cached_db_regions.go
index 00375928c58..33cb5ce5150 100644
--- a/pkg/sql/regions/cached_db_regions.go
+++ b/pkg/sql/regions/cached_db_regions.go
@@ -18,6 +18,7 @@ import (
     "github.com/cockroachdb/cockroach/pkg/roachpb"
     "github.com/cockroachdb/cockroach/pkg/sql/catalog"
     "github.com/cockroachdb/cockroach/pkg/sql/catalog/lease"
+    "github.com/cockroachdb/cockroach/pkg/util/hlc"
 )

 // CachedDatabaseRegions tracks the regions that are part of
@@ -30,11 +31,10 @@ type CachedDatabaseRegions struct {
 // NewCachedDatabaseRegions creates a new region cache by leasing
 // the underlying descriptors for the system database and region enum.
 func NewCachedDatabaseRegions(
-    ctx context.Context, db *kv.DB, lm *lease.Manager,
+    ctx context.Context, db *kv.DB, lm *lease.Manager, at hlc.Timestamp,
 ) (cdr *CachedDatabaseRegions, err error) {
     cdr = &CachedDatabaseRegions{}
-    now := db.Clock().Now()
-    desc, err := lm.Acquire(ctx, now, keys.SystemDatabaseID)
+    desc, err := lm.Acquire(ctx, at, keys.SystemDatabaseID)
     if err != nil {
         return nil, err
     }
@@ -48,7 +48,7 @@ func NewCachedDatabaseRegions(
         return nil, err
     }

-    desc, err = lm.Acquire(ctx, now, enumID)
+    desc, err = lm.Acquire(ctx, at, enumID)
     if err != nil {
         return nil, err
     }
```

that way we know that if `	systemDBVersion.AtLeast(clusterversion.V24_1_SessionBasedLeasingUpgradeDescriptor.Version())` returns true, that version came from a timestamp that's less than or equal to `at` in this function


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121070)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/regions/cached_db_regions.go` line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121070#-Ntw5b5K3o9YKf6unZwx:-Ntw5b5K3o9YKf6unZwy:baec7rb) ([raw file](https://github.com/cockroachdb/cockroach/blob/64ef1fd280933830f30b46a0acfac9a7e9288426/pkg/sql/regions/cached_db_regions.go#L33)):*
> ```Go
> // NewCachedDatabaseRegions creates a new region cache by leasing
> // the underlying descriptors for the system database and region enum.
> func NewCachedDatabaseRegions(
> ```

super nit: we could make a separate function called `NewCAchedDatabaseRegionsAt` and have the old implementation delegate to the new one. that way we don't have to update as many usages everywhere else


<!-- Sent from Reviewable.io -->

fqazi commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121070)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi)
> 
> _[`pkg/sql/regions/cached_db_regions.go` line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121070#-Ntw5b5K3o9YKf6unZwx:-Ntw5b5K3o9YKf6unZwy:baec7rb) ([raw file](https://github.com/cockroachdb/cockroach/blob/64ef1fd280933830f30b46a0acfac9a7e9288426/pkg/sql/regions/cached_db_regions.go#L33)):_
> 
> > ```go
> > // NewCachedDatabaseRegions creates a new region cache by leasing
> > // the underlying descriptors for the system database and region enum.
> > func NewCachedDatabaseRegions(
> > ```
> 
> super nit: we could make a separate function called `NewCAchedDatabaseRegionsAt` and have the old implementation delegate to the old one. that way we don't have to update as many usages everywhere else

Good point, let me do that

fqazi committed:
catalog/lease: use txn time when detecting session based leasing table

Previously, the logic to detect if the session based leasing table
should be injected synthetically queries the system database version,
but did not correctly use the transaction timestamp, which could lead to
us incorrectly assuming the descriptor is upgraded at the timestamp that
the leasing count query executed at. To address this, this patch updates
the CachedDatabaseRegions to support using a fixed timestamp from the
transaction.

Fixes: #120675, #121036

Release note: None

fqazi commented:
bors r+
